### PR TITLE
[PINOT-4603] Extended the protocol with the lease message

### PR DESCRIPTION
**Pull request by mcvsubbu**
Added protocol extension on the controller side to field the build-extension time
to be sent by the server.

The SegmentCompletionManager indicates to the server in the COMMIT response the max
time that the server can take before the SCM declares it dead. The SCM expects the
server to send a (newly introduced) extendBuildTime() message that extends the time
as often as needed.

Added test cases

### CODE REVIEW

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/SegmentCompletionManager.java [hunk 1]
```diff
@@ -481,6 +512,27 @@ public boolean isDone() {
...
+      final long now = _segmentCompletionManager.getCurrentTimeMs();
+      synchronized (this) {
+         LOGGER.info("Processing extendBuildTime({}, {}, {})", instanceId, offset, extTimeSec);
+        switch (_state) {
+          case PARTIAL_CONSUNIMG:
+          case HOLDING:
+          case COMMITTER_DECIDED:
+            return fail(instanceId, offset);
+          case COMMITTER_NOTIFIED:
+            return COMMITTER_NOTIFIED__extendBuildlTime(instanceId, offset, extTimeSec, now);
```
Comment (jfim) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/SegmentCompletionManager.java]: `BuildlTime`?

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/SegmentCompletionManager.java [hunk 2]
```diff
@@ -320,8 +345,19 @@ private SegmentCompletionFSM(PinotLLCRealtimeSegmentManager segmentManager,
...
+      long initialCommitTimeMs = MAX_TIME_TO_NOTIFY_WINNER_MS +
+              _segmentManager.getCommitTimeoutMS(_segmentName.getTableName());
+      Long savedCommitTime = _segmentCompletionManager._commitTimeMap.get(segmentName.getTableName());
+      if (savedCommitTime != null && savedCommitTime > initialCommitTimeMs) {
+        initialCommitTimeMs = savedCommitTime;
+      }
+      if (initialCommitTimeMs > MAX_COMMIT_TIME_FOR_ALL_SEGMENTS_SEC * 1000) {
+        // The table has a really high value configured for max commit time. Set it to a higher value than default
+        // and go from there.
+        initialCommitTimeMs = MAX_COMMIT_TIME_FOR_ALL_SEGMENTS_SEC * 1000;
```
Comment (jfim) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/SegmentCompletionManager.java]: We should log that this happened.

File: pinot-common/src/main/java/com/linkedin/pinot/common/protocols/SegmentCompletionProtocol.java [hunk 3]
```diff
@@ -102,21 +103,28 @@
   public static final String PARAM_OFFSET = "offset";
   public static final String PARAM_INSTANCE_ID = "instance";
   public static final String PARAM_REASON = "reason";
+  public static final String PARAM_EXT_TIME_SEC = "extTimeSec"; // Sent by servers to request additional time to build
```
Comment (mcvsubbu) [inline pinot-common/src/main/java/com/linkedin/pinot/common/protocols/SegmentCompletionProtocol.java]: changed to PARAM_EXTRA_TIME_SEC = "extraTimeSec"

File: pinot-common/src/main/java/com/linkedin/pinot/common/protocols/SegmentCompletionProtocol.java [hunk 4]
```diff
@@ -102,21 +103,28 @@
...
+  public static final String PARAM_NUM_ROWS = "nRows"; // Sent by servers to indicate the number of rows read so far
```
Comment (mcvsubbu) [inline pinot-common/src/main/java/com/linkedin/pinot/common/protocols/SegmentCompletionProtocol.java]: Changed to rowCount

File: pinot-common/src/main/java/com/linkedin/pinot/common/protocols/SegmentCompletionProtocol.java [hunk 5]
```diff
@@ -102,21 +103,28 @@
...
+  public static final String PARAM_BUILD_TIME_MS = "buildTimeMs"; // Time taken to build segment
+  public static final String PARAM_WAIT_TIME_MS = "waitTimeMs";   // Time taken to wait for build to start.
+
+  public static final String REASON_REACHED_MAX_ROWS = "maxRows";  // Stop reason sent by server as max num rows reached
```
Comment (mcvsubbu) [inline pinot-common/src/main/java/com/linkedin/pinot/common/protocols/SegmentCompletionProtocol.java]: changed to rowLimit and timeLimit

File: pinot-common/src/main/java/com/linkedin/pinot/common/protocols/SegmentCompletionProtocol.java [hunk 6]
```diff
@@ -270,6 +317,10 @@ public long getOffset() {
       return _offset;
     }
 
+    public long getBuildTimeSec() {
```
Comment (mcvsubbu) [inline pinot-common/src/main/java/com/linkedin/pinot/common/protocols/SegmentCompletionProtocol.java]: done

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/SegmentCompletionManager.java [hunk 7]
```diff
@@ -63,9 +63,16 @@
   private final HelixManager _helixManager;
   // A map that holds the FSM for each segment.
   private final Map<String, SegmentCompletionFSM> _fsmMap = new ConcurrentHashMap<>();
+  private final Map<String, Long> _commitTimeMap = new ConcurrentHashMap<>();
   private final PinotLLCRealtimeSegmentManager _segmentManager;
   private final ControllerMetrics _controllerMetrics;
 
+  private static final int MAX_COMMIT_TIME_FOR_ALL_SEGMENTS_SEC = 1800; // Half hour max commit time for all segments
```
Comment (mcvsubbu) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/SegmentCompletionManager.java]: done

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/SegmentCompletionManager.java (hunk 2 above)
Comment (mcvsubbu) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/SegmentCompletionManager.java]: done

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/SegmentCompletionManager.java [hunk 8]
```diff
@@ -752,6 +824,21 @@ private int numReplicasToLookFor() {
       return processStoppedConsuming(instanceId, offset, reason, false);
     }
 
+    private SegmentCompletionProtocol.Response COMMITTER_NOTIFIED__extendBuildlTime(String instanceId, long offset,
+        int extTimeSec, long now) {
+      SegmentCompletionProtocol.Response response = abortIfTooLateAndReturnHold(now, instanceId, offset);
+      if (response == null) {
+        long maxTimeAllowedToCommitMs = now + extTimeSec * 1000;
+        if (maxTimeAllowedToCommitMs > MAX_COMMIT_TIME_FOR_ALL_SEGMENTS_SEC * 1000) {
+          LOGGER.warn("Not accepting lease extension from {} requestedTime={}", _startTimeMs, maxTimeAllowedToCommitMs);
```
Comment (mcvsubbu) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/SegmentCompletionManager.java]: yes, done

**Pull request merged**